columnflow/selection/util.py

```diff
@@ -0,0 +1,68 @@
+# coding: utf-8
+
+"""
+Helpful utilities often used in selections.
+"""
+
+from __future__ import annotations
+
+__all__ = [
+    "sorted_indices_from_mask", "create_collections_from_masks",
+]
+
+from columnflow.util import maybe_import
+from columnflow.columnar_util import set_ak_column
+
+
+ak = maybe_import("awkward")
+
+
+def sorted_indices_from_mask(
+    mask: ak.Array,
+    metric: ak.Array,
+    sort_axis: int = -1,
+    ascending: bool = True,
+) -> ak.Array:
+    """
+    Takes a boolean *mask* and converts it to an array of indices, sorted using a *metric* of equal
+    size along a *sort_axis* and, by default, in *ascending* order. Example:
+
+    .. code-block:: python
+
+        mask = [[True, False, False, True], ...]
+        metric = [[5.0, 1.0, 0.9, 4.1], ...]
+
+        sorted_indices_from_mask(mask, metric)
+        # -> [[3, 0], ...]
+    """
+    indices = ak.argsort(metric, axis=sort_axis, ascending=ascending)
+    return indices[mask[indices]]
+
+
+def create_collections_from_masks(
+    events: ak.Array,
+    object_masks: dict[str, dict[str, ak.Array]] | ak.Array,
+    event_mask: ak.Array | type(Ellipsis) = Ellipsis,
```

Comment:
It looks like `event_mask` is the mask that has already been applied to `events`, but still needs to be applied to the subfields `object_masks` to make them index-compatible with `events`. I'm wondering if it wouldn't make more sense to require that `events` and `object_masks` be index-compatible already. That would and avoid the indirect dependence on `event_mask` and make the function behavior easier to describe.

Reply:
You're right of course. There should be no need to apply an event-level mask and filtering objects is also nothing I would expect from a function called `create_collections_from_masks`.

I removed the argument and moved the filtering to `ReduceEvents` where this method is used.